File: apps/mobile/android/app/src/main/java/com/tchat/services/AccessibilityManager.kt
/**
 * TchatAccessibilityManager.kt
 * TchatApp
 *
 * Created by Claude on 22/09/2024.
 */

package com.tchat.services

import android.accessibilityservice.AccessibilityServiceInfo
import android.content.Context
import android.content.res.Configuration
import android.graphics.Color
import android.os.Build
import android.provider.Settings
import android.util.Log
import android.view.View
import android.view.accessibility.AccessibilityManager
import android.view.accessibility.AccessibilityNodeInfo
import androidx.compose.foundation.layout.size
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.*
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow

/**
 * Accessibility compliance manager for Android
 * Implements T054: Accessibility compliance for TalkBack, Large Text, and WCAG guidelines
 */
class TchatAccessibilityManager private constructor(
    private val context: Context
) {

    // MARK: - Types

    enum class ComplianceLevel(val value: String) {
        A("A"),
        AA("AA"),
        AAA("AAA");

        val description: String
            get() = when (this) {
                A -> "WCAG 2.1 Level A (Basic)"
                AA -> "WCAG 2.1 Level AA (Standard)"
                AAA -> "WCAG 2.1 Level AAA (Enhanced)"
            }
    }

    data class AccessibilityValidationResult(
        val passes: Boolean,
        val actualRatio: Double,
        val requiredRatio: Double,
        val complianceLevel: ComplianceLevel,
        val message: String
    )

    data class AccessibilityRequirements(
        val colorContrastLevel: ComplianceLevel = ComplianceLevel.AA,
        val minimumTouchTargetSize: Float = 48f, // dp
        val requiresLabel: Boolean = true
    ) {
        companion object {
            val STANDARD = AccessibilityRequirements()
            val ENHANCED = AccessibilityRequirements(
                colorContrastLevel = ComplianceLevel.AAA,
                minimumTouchTargetSize = 56f
            )
        }
    }

    sealed class AccessibilityIssue(val description: String) {
        class MissingLabel(message: String) : AccessibilityIssue(message)
        class InsufficientTouchTarget(message: String) : AccessibilityIssue(message)
        class InsufficientContrast(message: String) : AccessibilityIssue(message)
        class MissingHint(message: String) : AccessibilityIssue(message)
        class IncorrectRole(message: String) : AccessibilityIssue(message)
    }

    // MARK: - Color Contrast Standards

    object ColorContrastRequirements {
        const val NORMAL_TEXT_AA = 4.5
        const val LARGE_TEXT_AA = 3.0
        const val NORMAL_TEXT_AAA = 7.0
        const val LARGE_TEXT_AAA = 4.5
        const val UI_COMPONENTS_AA = 3.0
    }

    // MARK: - State Properties

    private val accessibilityManager = context.getSystemService(Context.ACCESSIBILITY_SERVICE) as android.view.accessibility.AccessibilityManager

    private val _isTalkBackEnabled = MutableStateFlow(false)
    val isTalkBackEnabled: StateFlow<Boolean> = _isTalkBackEnabled.asStateFlow()

    private val _fontScale = MutableStateFlow(1f)
    val fontScale: StateFlow<Float> = _fontScale.asStateFlow()

    private val _isHighContrastEnabled = MutableStateFlow(false)
    val isHighContrastEnabled: StateFlow<Boolean> = _isHighContrastEnabled.asStateFlow()

    private val _isReduceMotionEnabled = MutableStateFlow(false)
    val isReduceMotionEnabled: StateFlow<Boolean> = _isReduceMotionEnabled.asStateFlow()

    companion object {
        private const val TAG = "AccessibilityManager"

        @Volatile
        private var INSTANCE: TchatAccessibilityManager? = null

        fun getInstance(context: Context): TchatAccessibilityManager {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: TchatAccessibilityManager(context.applicationContext).also { INSTANCE = it }
            }
        }
    }

    // MARK: - Public Interface

    /**
     * Starts monitoring accessibility changes
     */
    fun startMonitoring() {
        updateAccessibilitySettings()
        setupAccessibilityListener()
    }

    /**
     * Stops monitoring accessibility changes
     */
    fun stopMonitoring() {
        // No specific cleanup needed for Android accessibility monitoring
    }

    /**
     * Validates color contrast ratio according to WCAG guidelines
     */
    fun validateColorContrast(
        foregroundColor: Int,
        backgroundColor: Int,
        fontSize: Float, // sp
        complianceLevel: ComplianceLevel = ComplianceLevel.AA
    ): AccessibilityValidationResult {

        val contrastRatio = calculateContrastRatio(foregroundColor, backgroundColor)
        val isLargeText = fontSize >= 18f || (fontSize >= 14f && isBoldText(fontSize))

        val requiredRatio = when (complianceLevel) {
            ComplianceLevel.A -> 3.0 // Basic requirement
            ComplianceLevel.AA -> if (isLargeText) ColorContrastRequirements.LARGE_TEXT_AA else ColorContrastRequirements.NORMAL_TEXT_AA
            ComplianceLevel.AAA -> if (isLargeText) ColorContrastRequirements.LARGE_TEXT_AAA else ColorContrastRequirements.NORMAL_TEXT_AAA
        }

        val passes = contrastRatio >= requiredRatio

        return AccessibilityValidationResult(
            passes = passes,
            actualRatio = contrastRatio,
            requiredRatio = requiredRatio,
            complianceLevel = complianceLevel,
            message = if (passes) {
                "Color contrast meets ${complianceLevel.description}"
            } else {
                "Color contrast (${String.format("%.2f", contrastRatio)}:1) below required ${String.format("%.1f", requiredRatio)}:1 for ${complianceLevel.description}"
            }
        )
    }

    /**
     * Configures accessibility for Android Views
     */
    fun configureAccessibility(
        view: View,
        contentDescription: String,
        hint: String? = null,
        role: AccessibilityNodeInfo.AccessibilityAction? = null
    ) {
        view.contentDescription = contentDescription
        view.importantForAccessibility = View.IMPORTANT_FOR_ACCESSIBILITY_YES

        hint?.let {
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
                view.tooltipText = it
            }
        }

        // Configure accessibility actions if provided
        role?.let { action ->
            view.setAccessibilityDelegate(object : View.AccessibilityDelegate() {
                override fun onInitializeAccessibilityNodeInfo(host: View, info: AccessibilityNodeInfo) {
                    super.onInitializeAccessibilityNodeInfo(host, info)
                    info.addAction(action)
                }
            })
        }
    }

    /**
     * Announces accessibility changes to TalkBack
     */
    fun announceChange(message: String) {
        if (_isTalkBackEnabled.value) {
            // Create a temporary view to announce the message
            val view = View(context)
            view.announceForAccessibility(message)
        }
    }

    /**
     * Gets scaled font size based on system font scale
     */
    fun getScaledTextSize(baseSize: Float): Float {
        return baseSize * _fontScale.value
    }

    /**
     * Checks if device settings require reduced motion
     */
    val shouldReduceMotion: Boolean
        get() = _isReduceMotionEnabled.value

    /**
     * Gets appropriate animation duration based on accessibility settings
     */
    fun getAnimationDuration(defaultDuration: Long): Long {
        return if (shouldReduceMotion) 0L else defaultDuration
    }

    /**
     * Validates component accessibility compliance
     */
    fun validateComponentAccessibility(
        view: View,
        requirements: AccessibilityRequirements = AccessibilityRequirements.STANDARD
    ): List<AccessibilityIssue> {
        val issues = mutableListOf<AccessibilityIssue>()

        // Check for content description
        if (requirements.requiresLabel && view.contentDescription.isNullOrEmpty()) {
            issues.add(AccessibilityIssue.MissingLabel("Component missing content description"))
        }

        // Check minimum touch target size (convert dp to pixels)
        val density = context.resources.displayMetrics.density
        val minSizePx = requirements.minimumTouchTargetSize * density

        if (view.width < minSizePx || view.height < minSizePx) {
            issues.add(AccessibilityIssue.InsufficientTouchTarget(
                "Touch target smaller than ${requirements.minimumTouchTargetSize}dp"
            ))
        }

        return issues
    }

    /**
     * Provides minimum touch target size for accessibility compliance
     */
    fun getMinimumTouchTarget(): Dp = 48.dp

    // MARK: - Private Methods

    private fun setupAccessibilityListener() {
        accessibilityManager.addAccessibilityStateChangeListener { enabled ->
            updateAccessibilitySettings()
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            accessibilityManager.addTouchExplorationStateChangeListener { enabled ->
                updateAccessibilitySettings()
            }
        }
    }

    private fun updateAccessibilitySettings() {
        updateTalkBackStatus()
        updateFontScale()
        updateHighContrastStatus()
        updateReduceMotionStatus()
    }

    private fun updateTalkBackStatus() {
        val enabledServices = accessibilityManager.getEnabledAccessibilityServiceList(AccessibilityServiceInfo.FEEDBACK_SPOKEN)
        _isTalkBackEnabled.value = enabledServices.isNotEmpty()
    }

    private fun updateFontScale() {
        _fontScale.value = context.resources.configuration.fontScale
    }

    private fun updateHighContrastStatus() {
        _isHighContrastEnabled.value = try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                Settings.Secure.getFloat(
                    context.contentResolver,
                    "high_text_contrast_enabled",
                    0f
                ) == 1f
            } else {
                false
            }
        } catch (e: Settings.SettingNotFoundException) {
            false
        }
    }

    private fun updateReduceMotionStatus() {
        _isReduceMotionEnabled.value = try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                Settings.Global.getFloat(
                    context.contentResolver,
                    Settings.Global.ANIMATOR_DURATION_SCALE,
                    1f
                ) == 0f
            } else {
                false
            }
        } catch (e: Settings.SettingNotFoundException) {
            false
        }
    }

    // MARK: - Color Contrast Calculations

    private fun calculateContrastRatio(foregroundColor: Int, backgroundColor: Int): Double {
        val foregroundLuminance = relativeLuminance(foregroundColor)
        val backgroundLuminance = relativeLuminance(backgroundColor)

        val lighter = max(foregroundLuminance, backgroundLuminance)
        val darker = min(foregroundLuminance, backgroundLuminance)

        return (lighter + 0.05) / (darker + 0.05)
    }

    private fun relativeLuminance(color: Int): Double {
        val red = Color.red(color) / 255.0
        val green = Color.green(color) / 255.0
        val blue = Color.blue(color) / 255.0

        // Convert to linear RGB
        val linearRed = linearizeColorComponent(red)
        val linearGreen = linearizeColorComponent(green)
        val linearBlue = linearizeColorComponent(blue)

        // Calculate relative luminance using ITU-R BT.709 coefficients
        return 0.2126 * linearRed + 0.7152 * linearGreen + 0.0722 * linearBlue
    }

    private fun linearizeColorComponent(component: Double): Double {
        return if (component <= 0.03928) {
            component / 12.92
        } else {
            ((component + 0.055) / 1.055).pow(2.4)
        }
    }

    // MARK: - Helper Methods

    private fun isBoldText(fontSize: Float): Boolean {
        // Simplified bold text detection
        return fontSize >= 14f
    }
}

// MARK: - Compose Extensions

/**
 * Modifier for accessibility compliance in Jetpack Compose
 */
fun Modifier.accessibilityCompliance(
    contentDescription: String,
    role: Role? = null,
    stateDescription: String? = null,
    onClick: (() -> Unit)? = null
): Modifier = this.semantics {
    this.contentDescription = contentDescription
    role?.let { this.role = it }
    stateDescription?.let { this.stateDescription = it }
    onClick?.let { this.onClick { it(); true } }
}

/**
 * Modifier for minimum touch target size
 */
fun Modifier.minimumTouchTarget(size: Dp = 48.dp): Modifier = this.size(size)

/**
 * Get scaled text size for accessibility
 */
@Composable
fun getAccessibleTextSize(baseSize: Float): androidx.compose.ui.unit.TextUnit {
    val context = LocalContext.current
    val configuration = LocalConfiguration.current
    val scaledSize = baseSize * configuration.fontScale
    return scaledSize.sp
}

/**
 * Check if high contrast is enabled
 */
@Composable
fun isHighContrastEnabled(): Boolean {
    val context = LocalContext.current
    return try {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            Settings.Secure.getFloat(
                context.contentResolver,
                "high_text_contrast_enabled",
                0f
            ) == 1f
        } else {
            false
        }
    } catch (e: Settings.SettingNotFoundException) {
        false
    }
}

/**
 * Accessibility-aware color selection
 */
@Composable
fun accessibleColor(
    normal: androidx.compose.ui.graphics.Color,
    highContrast: androidx.compose.ui.graphics.Color
): androidx.compose.ui.graphics.Color {
    return if (isHighContrastEnabled()) highContrast else normal
}

File: apps/mobile/ios/Sources/Services/AccessibilityManager.swift
//
//  AccessibilityManager.swift
//  TchatApp
//
//  Created by Claude on 22/09/2024.
//

import Foundation
#if canImport(UIKit)
import UIKit
#endif
import SwiftUI

/// Accessibility compliance manager for iOS
/// Implements T054: Accessibility compliance for VoiceOver, Dynamic Type, and WCAG guidelines
public class AccessibilityManager: ObservableObject {

    // MARK: - Published Properties

    @Published public var isVoiceOverRunning: Bool = false
    @Published public var preferredContentSizeCategory: ContentSizeCategory = .medium
    @Published public var isReduceMotionEnabled: Bool = false
    @Published public var isHighContrastEnabled: Bool = false
    @Published public var isDifferentiateWithoutColorEnabled: Bool = false

    // MARK: - Private Properties

    private var notificationCenter = NotificationCenter.default
    private var observers: [NSObjectProtocol] = []

    // MARK: - Accessibility Compliance Levels

    public enum ComplianceLevel: String, CaseIterable {
        case a = "A"
        case aa = "AA"
        case aaa = "AAA"

        var description: String {
            switch self {
            case .a: return "WCAG 2.1 Level A (Basic)"
            case .aa: return "WCAG 2.1 Level AA (Standard)"
            case .aaa: return "WCAG 2.1 Level AAA (Enhanced)"
            }
        }
    }

    // MARK: - Color Contrast Standards

    public struct ColorContrastRequirements {
        public static let normalTextAA: Double = 4.5
        public static let largeTextAA: Double = 3.0
        public static let normalTextAAA: Double = 7.0
        public static let largeTextAAA: Double = 4.5
        public static let uiComponentsAA: Double = 3.0
    }

    // MARK: - Initialization

    public static let shared = AccessibilityManager()

    private init() {
        setupNotificationObservers()
        updateAccessibilitySettings()
    }

    deinit {
        removeNotificationObservers()
    }

    // MARK: - Public Interface

    /// Starts monitoring accessibility changes
    public func startMonitoring() {
        updateAccessibilitySettings()
    }

    /// Stops monitoring accessibility changes
    public func stopMonitoring() {
        removeNotificationObservers()
    }

    /// Validates color contrast ratio according to WCAG guidelines
    public func validateColorContrast(
        foreground: UIColor,
        background: UIColor,
        fontSize: CGFloat,
        complianceLevel: ComplianceLevel = .aa
    ) -> AccessibilityValidationResult {

        let contrastRatio = calculateContrastRatio(foreground: foreground, background: background)
        let isLargeText = fontSize >= 18.0 || (fontSize >= 14.0 && isBoldText(fontSize))

        let requiredRatio: Double
        switch complianceLevel {
        case .a:
            requiredRatio = 3.0 // Basic requirement
        case .aa:
            requiredRatio = isLargeText ? ColorContrastRequirements.largeTextAA : ColorContrastRequirements.normalTextAA
        case .aaa:
            requiredRatio = isLargeText ? ColorContrastRequirements.largeTextAAA : ColorContrastRequirements.normalTextAAA
        }

        let passes = contrastRatio >= requiredRatio

        return AccessibilityValidationResult(
            passes: passes,
            actualRatio: contrastRatio,
            requiredRatio: requiredRatio,
            complianceLevel: complianceLevel,
            message: passes ? "Color contrast meets \(complianceLevel.description)" :
                    "Color contrast (\(String(format: "%.2f", contrastRatio)):1) below required \(String(format: "%.1f", requiredRatio)):1 for \(complianceLevel.description)"
        )
    }

    /// Configures accessibility for UI elements
    public func configureAccessibility(
        for view: UIView,
        label: String,
        hint: String? = nil,
        traits: UIAccessibilityTraits = [],
        value: String? = nil
    ) {
        view.isAccessibilityElement = true
        view.accessibilityLabel = label
        view.accessibilityHint = hint
        view.accessibilityTraits = traits
        view.accessibilityValue = value
    }

    /// Configures accessibility for SwiftUI views
    public func accessibilityModifier(
        label: String,
        hint: String? = nil,
        value: String? = nil,
        traits: AccessibilityTraits = []
    ) -> some ViewModifier {
        return AccessibilityConfigurationModifier(
            label: label,
            hint: hint,
            value: value,
            traits: traits
        )
    }

    /// Announces accessibility changes to VoiceOver
    public func announceChange(_ message: String, priority: UIAccessibility.Notification = .announcement) {
        if isVoiceOverRunning {
            UIAccessibility.post(notification: priority, argument: message)
        }
    }

    /// Gets optimal font size for current Dynamic Type setting
    public func scaledFont(baseSize: CGFloat, textStyle: UIFont.TextStyle = .body) -> UIFont {
        let font = UIFont.systemFont(ofSize: baseSize)
        return UIFontMetrics(forTextStyle: textStyle).scaledFont(for: font)
    }

    /// Checks if device settings require reduced motion
    public var shouldReduceMotion: Bool {
        return isReduceMotionEnabled
    }

    /// Gets appropriate animation duration based on accessibility settings
    public func animationDuration(default duration: TimeInterval) -> TimeInterval {
        return shouldReduceMotion ? 0.0 : duration
    }

    /// Validates component accessibility compliance
    public func validateComponentAccessibility(
        component: UIView,
        requirements: AccessibilityRequirements = .standard
    ) -> [AccessibilityIssue] {
        var issues: [AccessibilityIssue] = []

        // Check for accessibility label
        if component.isAccessibilityElement && component.accessibilityLabel?.isEmpty != false {
            issues.append(.missingLabel("Component missing accessibility label"))
        }

        // Check minimum touch target size (44x44 points for AA compliance)
        if component.frame.width < 44 || component.frame.height < 44 {
            issues.append(.insufficientTouchTarget("Touch target smaller than 44x44 points"))
        }

        // Check for color contrast if applicable
        if let backgroundColor = component.backgroundColor,
           let textColor = getTextColor(from: component) {
            let validation = validateColorContrast(
                foreground: textColor,
                background: backgroundColor,
                fontSize: getFontSize(from: component),
                complianceLevel: requirements.colorContrastLevel
            )

            if !validation.passes {
                issues.append(.insufficientContrast(validation.message))
            }
        }

        return issues
    }

    // MARK: - Private Methods

    private func setupNotificationObservers() {
        // VoiceOver status changes
        observers.append(
            notificationCenter.addObserver(
                forName: UIAccessibility.voiceOverStatusDidChangeNotification,
                object: nil,
                queue: .main
            ) { [weak self] _ in
                self?.updateVoiceOverStatus()
            }
        )

        // Dynamic Type changes
        observers.append(
            notificationCenter.addObserver(
                forName: UIContentSizeCategory.didChangeNotification,
                object: nil,
                queue: .main
            ) { [weak self] _ in
                self?.updateContentSizeCategory()
            }
        )

        // Reduce Motion changes
        observers.append(
            notificationCenter.addObserver(
                forName: UIAccessibility.reduceMotionStatusDidChangeNotification,
                object: nil,
                queue: .main
            ) { [weak self] _ in
                self?.updateReduceMotionStatus()
            }
        )

        // High Contrast changes
        observers.append(
            notificationCenter.addObserver(
                forName: UIAccessibility.darkerSystemColorsStatusDidChangeNotification,
                object: nil,
                queue: .main
            ) { [weak self] _ in
                self?.updateHighContrastStatus()
            }
        )

        // Differentiate Without Color changes
        observers.append(
            notificationCenter.addObserver(
                forName: UIAccessibility.differentiateWithoutColorDidChangeNotification,
                object: nil,
                queue: .main
            ) { [weak self] _ in
                self?.updateDifferentiateWithoutColorStatus()
            }
        )
    }

    private func removeNotificationObservers() {
        observers.forEach { notificationCenter.removeObserver($0) }
        observers.removeAll()
    }

    private func updateAccessibilitySettings() {
        updateVoiceOverStatus()
        updateContentSizeCategory()
        updateReduceMotionStatus()
        updateHighContrastStatus()
        updateDifferentiateWithoutColorStatus()
    }

    private func updateVoiceOverStatus() {
        isVoiceOverRunning = UIAccessibility.isVoiceOverRunning
    }

    private func updateContentSizeCategory() {
        let category = UIApplication.shared.preferredContentSizeCategory
        preferredContentSizeCategory = ContentSizeCategory(from: category)
    }

    private func updateReduceMotionStatus() {
        isReduceMotionEnabled = UIAccessibility.isReduceMotionEnabled
    }

    private func updateHighContrastStatus() {
        isHighContrastEnabled = UIAccessibility.isDarkerSystemColorsEnabled
    }

    private func updateDifferentiateWithoutColorStatus() {
        isDifferentiateWithoutColorEnabled = UIAccessibility.shouldDifferentiateWithoutColor
    }

    // MARK: - Color Contrast Calculations

    private func calculateContrastRatio(foreground: UIColor, background: UIColor) -> Double {
        let foregroundLuminance = relativeLuminance(of: foreground)
        let backgroundLuminance = relativeLuminance(of: background)

        let lighter = max(foregroundLuminance, backgroundLuminance)
        let darker = min(foregroundLuminance, backgroundLuminance)

        return (lighter + 0.05) / (darker + 0.05)
    }

    private func relativeLuminance(of color: UIColor) -> Double {
        var red: CGFloat = 0
        var green: CGFloat = 0
        var blue: CGFloat = 0
        var alpha: CGFloat = 0

        color.getRed(&red, green: &green, blue: &blue, alpha: &alpha)

        // Convert to linear RGB
        let linearRed = linearizeColorComponent(Double(red))
        let linearGreen = linearizeColorComponent(Double(green))
        let linearBlue = linearizeColorComponent(Double(blue))

        // Calculate relative luminance using ITU-R BT.709 coefficients
        return 0.2126 * linearRed + 0.7152 * linearGreen + 0.0722 * linearBlue
    }

    private func linearizeColorComponent(_ component: Double) -> Double {
        if component <= 0.03928 {
            return component / 12.92
        } else {
            return pow((component + 0.055) / 1.055, 2.4)
        }
    }

    // MARK: - Helper Methods

    private func isBoldText(_ fontSize: CGFloat) -> Bool {
        // Simplified bold text detection
        return fontSize >= 14.0
    }

    private func getTextColor(from view: UIView) -> UIColor? {
        if let label = view as? UILabel {
            return label.textColor
        } else if let button = view as? UIButton {
            return button.titleColor(for: .normal)
        }
        return nil
    }

    private func getFontSize(from view: UIView) -> CGFloat {
        if let label = view as? UILabel {
            return label.font.pointSize
        } else if let button = view as? UIButton {
            return button.titleLabel?.font.pointSize ?? 17.0
        }
        return 17.0 // Default system font size
    }
}

// MARK: - Supporting Types

public struct AccessibilityValidationResult {
    public let passes: Bool
    public let actualRatio: Double
    public let requiredRatio: Double
    public let complianceLevel: AccessibilityManager.ComplianceLevel
    public let message: String
}

public struct AccessibilityRequirements {
    public let colorContrastLevel: AccessibilityManager.ComplianceLevel
    public let minimumTouchTargetSize: CGSize
    public let requiresLabel: Bool

    public static let standard = AccessibilityRequirements(
        colorContrastLevel: .aa,
        minimumTouchTargetSize: CGSize(width: 44, height: 44),
        requiresLabel: true
    )

    public static let enhanced = AccessibilityRequirements(
        colorContrastLevel: .aaa,
        minimumTouchTargetSize: CGSize(width: 48, height: 48),
        requiresLabel: true
    )
}

public enum AccessibilityIssue {
    case missingLabel(String)
    case insufficientTouchTarget(String)
    case insufficientContrast(String)
    case missingHint(String)
    case incorrectTraits(String)

    public var description: String {
        switch self {
        case .missingLabel(let message),
             .insufficientTouchTarget(let message),
             .insufficientContrast(let message),
             .missingHint(let message),
             .incorrectTraits(let message):
            return message
        }
    }
}

// MARK: - SwiftUI Extensions

public extension ContentSizeCategory {
    init(from uiContentSizeCategory: UIContentSizeCategory) {
        switch uiContentSizeCategory {
        case .extraSmall: self = .extraSmall
        case .small: self = .small
        case .medium: self = .medium
        case .large: self = .large
        case .extraLarge: self = .extraLarge
        case .extraExtraLarge: self = .extraExtraLarge
        case .extraExtraExtraLarge: self = .extraExtraExtraLarge
        case .accessibilityMedium: self = .accessibilityMedium
        case .accessibilityLarge: self = .accessibilityLarge
        case .accessibilityExtraLarge: self = .accessibilityExtraLarge
        case .accessibilityExtraExtraLarge: self = .accessibilityExtraExtraLarge
        case .accessibilityExtraExtraExtraLarge: self = .accessibilityExtraExtraExtraLarge
        default: self = .medium
        }
    }
}

struct AccessibilityConfigurationModifier: ViewModifier {
    let label: String
    let hint: String?
    let value: String?
    let traits: AccessibilityTraits

    func body(content: Content) -> some View {
        content
            .accessibilityLabel(label)
            .accessibilityHint(hint ?? "")
            .accessibilityValue(value ?? "")
            .accessibilityAddTraits(traits)
    }
}

// MARK: - View Extensions

public extension View {
    func accessibilityConfiguration(
        label: String,
        hint: String? = nil,
        value: String? = nil,
        traits: AccessibilityTraits = []
    ) -> some View {
        modifier(AccessibilityConfigurationModifier(
            label: label,
            hint: hint,
            value: value,
            traits: traits
        ))
    }

    func dynamicTypeSize(min: DynamicTypeSize = .xSmall, max: DynamicTypeSize = .accessibility5) -> some View {
        self.dynamicTypeSize(min...max)
    }
}